azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -228,23 +366,75 @@ def __purge_extra_agents_from_disk(current_version, known_agents):
             except Exception as e:
                 logger.warn(u"Purging {0} raised exception: {1}", agent_path, ustr(e))
 
-    def __proceed_with_update(self, requested_version):
+
+class RSMVersionUpdater(GAVersionUpdater):
+    def __init__(self, gs_id, agent_family, agent_manifest, version, update_state, is_version_from_rsm, daemon_version):
+        super(RSMVersionUpdater, self).__init__(gs_id, agent_family, agent_manifest, version, update_state)
+        self._is_version_from_rsm = is_version_from_rsm
+        self._daemon_version = daemon_version
+
+    @staticmethod
+    def _get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if os.path.isdir(agent_dir)]
+
+    def _get_available_agents_on_disk(self):
+        available_agents = [agent for agent in self._get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    def _is_update_allowed_this_time(self):
         """
-        If requested version is specified, upgrade/downgrade to the specified version.
-        Raises: AgentUpgradeExitException
+        update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+        If update allowed, we update the last_attempted_rsm_version_update_time to current time.
         """
-        if requested_version < CURRENT_VERSION:
+        now = datetime.datetime.now()
+
+        if self._update_state.last_attempted_rsm_version_update_time != datetime.datetime.min:
+            next_attempt_time = self._update_state.last_attempted_rsm_version_update_time + datetime.timedelta(
+                seconds=conf.get_autoupdate_frequency())
+        else:
+            next_attempt_time = now
+
+        if next_attempt_time > now:
+            return False
+        self._update_state.last_attempted_rsm_version_update_time = now
+        # The time limit elapsed for us to allow updates.
+        return True
+
+    def should_update_agent(self, goal_state):
+        if not self._is_update_allowed_this_time():
+            return False
+
+        # we don't allow updates if version is not from RSM or downgrades below daemon version or if version is same as current version
+        if not self._is_version_from_rsm or self._version < self._daemon_version or self._version == CURRENT_VERSION:
+            return False
```

Comment:
If we hit this case then we don't end up doing the update, but we updated our attempt time to be now already in _is_update_allowed_this_time (line 400).

Is this the behavior we want? I think we shouldn't update the 'last_attempted_rsm_version_update_time' until we actually attempt the update

Reply:
That is debatable. What I thought was, if we don't update the clock, if rsm request made to vms same time, then we end up does the update to all vms at the time. Even though this is small window, still does some help to spread.

Reply:
I added a comment similar to @maddieford 's in a previous iteration of this code. I also find that confusing